### PR TITLE
feat(monitoring): Introduced monitoring documentation (incomplete).

### PR DESCRIPTION
**Pull request by ewiseblatt**
@ttomsu 

This does not yet include datadog, but is a stopping point for now.

### CODE REVIEW

File: setup/monitoring/index.md [hunk 1]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+that it should be straight forward to add other systems as well. In fact,
+each of the supported systems was provided using the daemon's extension
+mechanisms -- there are not "native" systems.
+
+You can also use the microservice HTTP endpoint `/spectator/metrics`
+directly to scrape metrics yourself. The JSON document structure is 
+further documented in the [Monitoring reference section](/reference/monitoring/).
+
+
+## Installing Spinnaker Monitoring (on VMs)
```
Comment (lwander) [inline setup/monitoring/index.md]: Let's talk tomorrow in person about how to split divide this work w.r.t. work done by halyard vs work done by the user.

File: reference/monitoring/index.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+## Enabling Monitoring
+
+The monitoring endpoint, `/spectator/metrics`, is not available unless
+monitoring is enabled, which it is by default. Monitoring is enabled in
+the microservices using the Halyard command **TBD**.
+
+
+## Metrics Overview
+
+Metrics are recorded with a set of tag bindings (e.g. `success=true`) describing
```
Comment (ttomsu) [inline reference/monitoring/index.md]: While this first paragraph is information dense, it's hard to parse for a new reader. I'd suggest something like:

"Each metric that is collected has two parts: tags and values. Tags describe and give context to the value. For example, a web request is handled by a specific _controller_, and each controller can emits one or more metrics. Lets look at an example metric:"

File: reference/monitoring/index.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+"values": [
+    {
+        "t": 1493930053715,
+        "v": 6450.0
+    }
+]
+```
+
+The interpretation of this is that as of time `t=1493930053715` (milliseconds
+since unix epoch), there were `v=6450` calls to the `listByApplication
```
Comment (ttomsu) [inline reference/monitoring/index.md]: Simplify the examples a bit - "t=1500000000000" and "v=100"

Comment (ttomsu) [reply reference/monitoring/index.md]: Also missing a trailing tick after `listByApplication` - it's causing formatting issues.

File: setup/monitoring/index.md [hunk 4]
```diff
@@ -6,3 +6,262 @@ sidebar:
 ---
 
 {% include toc %}
+Spinnaker is instrumented with numerous metrics internally so that it can be
+monitored. Monitoring spinnaker typically involves the spinnaker-monitoring
+daemon, awhich collects metrics reported by each microservice instance and
```
Comment (ttomsu) [inline setup/monitoring/index.md]: typo: "awhich"

File: setup/monitoring/index.md [hunk 5]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+overview dashboards, receive alerts, and informally browse depending on your
+needs.
+
+Spinnaker publishes internal metrics using a multi-dimensional data model
+based on "tags". The metrics, data-model, and usage a discussed further
+in the sections [Consuming Metrics](#consuming-metrics) and in the
+[Monitoring Reference document](/reference/monitoring/).
+
+Spinnaker currently supports three specific third-party systems: 
+[Datadog](https://www.datadoghq.com/), [Prometheus](https://prometheus.io/)
```
Comment (ttomsu) [inline setup/monitoring/index.md]: Put Prometheus first, since it's the best supported?

File: setup/monitoring/index.md [hunk 6]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+
+Spinnaker publishes internal metrics using a multi-dimensional data model
+based on "tags". Each "metric" has a name and type. Each data point is a
+numeric value that is time-stamped at the time of reporting and tagged with
+a set of one or more "label"="value" tags. These tag values are strings,
+though some may have numeric-looking values. Taken together, the set of
+tags convey the context for the reported measurement. Each of these
+contexts forms a distinct time-series data stream.
+
+For example a metric counting calls may be tagged with a "status" label
```
Comment (ttomsu) [inline setup/monitoring/index.md]: counting web requests?

File: setup/monitoring/index.md [hunk 7]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+and values indicating whether the call was successful or not. So rather
+than having two metrics, one for successful calls and the other for
+unsuccessful calls, there is a single metric, where the underlying
+monitoring system can filter the successful from unsuccessful as you want
+depending on how you wish to abstract and interpret the data. In practice
+the metrics have many tags providing a lot of granularity and ways in
+which you can aggregate and interpret them. The data model is described
+further in [the Monitoring reference section](/reference/monitoring/).
```
Comment (ttomsu) [inline setup/monitoring/index.md:92]: It's probably not necessary to link to the reference section 3+ times.

File: setup/monitoring/index.md [hunk 8]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+     * [Configuring Datadog](/setup/monitoring/datadog)
+     * [Configuring Prometheus](/setup/monitoring/prometheus)
+     * [Configuring Stackdriver](/setup/monitoring/stackdriver)
+
+  4. Restart the Spinnaker Monitoring Daemon
+```
+sudo service spinnaker-monitoring restart
+```
+
+# Configuring the Spinnaker Monitoring Daemon
```
Comment (ttomsu) [inline setup/monitoring/index.md]: On the final page, the title is usually the only thing that gets the `h1` header. I suggest indenting each header with one more hash

File: setup/monitoring/index.md [hunk 9]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+microservice they came from (how it does so depends on the backend
+metrics system you are using). There are no global "Spinnaker" metrics.
+There are only individual metrics on the individual instance of
+concrete microservices that, taken together, compose a "Spinnaker" deployment.
+
+
+### Types of Metrics
+
+There are two basic types of metrics currently supported,
+*counters* and *gauges*.
```
Comment (ttomsu) [inline setup/monitoring/index.md:116]: I like that you've defined these out. Maybe in the reference section point back to this page to understand the concepts?

File: setup/monitoring/index.md [hunk 10]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+    those counters will be independent of one another. Each process only
+    knows about itself. The daemon adds a tag to each data point that
+    identifies which instance it came from so that you can drill down
+    into individual instances if you need. However, typically you will
+    use your monitoring system to aggregate counters across all replicas.
+
+    Counters are useful to determine rates. Given two points in time,
+    the counter differences will be the measurement delta and the
+    delta divided by the time difference will be the rate.
+    (divide by another 1000000 to convert nanoseconds to milliseconds).
```
Comment (ttomsu) [inline setup/monitoring/index.md]: Is a rate at item per millisecond useful? I've only seen metrics measured at item per second.

File: setup/monitoring/index.md [hunk 11]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+
+    * Spinnaker has a special type of counter called a *Timer*.
```
Comment (ttomsu) [inline setup/monitoring/index.md]: (From a newcomer's perspective): Isn't this all Spinnaker?

My guess is that you're separating out what _Spectator_ can do and what Spinnaker builds on top of that, but to a reader I don't think that distinction is relevant. However, I may be interpreting this wrong.

File: setup/monitoring/index.md [hunk 12]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+
+      The Monitoring Daemon transforms these into two distinct
+      counters and removes the "statistic" label entirely. The
+      "count" counter is named with the original metric name plus
+      a "__count" suffix. The "totalTime" with a "__totalTime" suffix.
+
+      The "count" represents the number of measurements taken.
+      The "totalTime" represents the number of nanoseconds measured
+      across all the calls. Dividing the "totalTime" by the "count"
+      over some time window gives the latency over that time window.
```
Comment (ttomsu) [inline setup/monitoring/index.md:155]: A simple example would be good to illustrate this concept.

File: setup/monitoring/index.md [hunk 13]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+      __Timers__ are used to measure timing information. These are
+      always in nanoseconds. When consuming metrics straight from
+      Spinnaker, a Timer will have two complementary time series.
+      One will have a tag "statistic" with the value "count" and
+      the other a tag with a "statistic" with the value "totalTime".
+
+      The Monitoring Daemon transforms these into two distinct
+      counters and removes the "statistic" label entirely. The
+      "count" counter is named with the original metric name plus
+      a "__count" suffix. The "totalTime" with a "__totalTime" suffix.
```
Comment (ttomsu) [inline setup/monitoring/index.md:150]: If the fact that "statistic" is not relevant to the end user, I wouldn't even mention it. You can add this bit to the Reference section and mention that the daemon doesn't use it for anything, so anyone writing their own direct scraping impl knows they can drop it too.

File: setup/monitoring/index.md [hunk 14]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+"statusCode", and "success". Some microservices may add an additional
+label such as "account" or "application" depending on the nature of
+the microservices API.
+
+These metrics will have several time series, such as those with the
+following tag bindings:
+
+account | controller | method | status | statusCode | success
+-----------|--------------|------------|----------|-----------------|-----
+my-aws-account | ClusterController | getForAccountAndNameAndType | 2xx | 200 | true
```
Comment (ttomsu) [inline setup/monitoring/index.md]: You could probably just use `my-account` to keep it provider agnostic

File: setup/monitoring/index.md [hunk 15]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+
+Each microservice has a `controller.invocations` metric used to
+instrument API calls into it. Since this is a timer, in practice
+this is broken out into two 'controller.invocations\_\_count' and
+'controller.invocations\_\_totalTime'.
+
+These typically have the labels "controller", "method", "status",
+"statusCode", and "success". Some microservices may add an additional
+label such as "account" or "application" depending on the nature of
+the microservices API.
```
Comment (ttomsu) [inline setup/monitoring/index.md:196]: Maybe mention that this example is from an actual controller on Gate?

File: setup/monitoring/index.md [hunk 16]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+semantics. Some of these tags may be of more interest than others. In
+the case above, some of the tags are at different levels of abstraction
+and not actually independent. For example a 2xx status will always be
+success=true and a non-2xx status code will always be success=false.
+Which to use is a matter of convienence but given the status tag (which
+can distinguish 4xx from 5xx errors) the success tag does not add any
+additional time-series permutations since its value is not actually
+independent.
+
+### About the Supplied Dashboards
```
Comment (ttomsu) [inline setup/monitoring/index.md]: "Pre-configured Dashboards" ? "Pre-built Dashboards" ? Or just simply "Supplied Dashboards"

File: setup/monitoring/index.md [hunk 17]
```diff
@@ -6,3 +6,262 @@ sidebar:
...
+(over rates) for 1-minute sliding windows. This might change in the future.
+Some of the caveats here are due to the choice to show values over rates, but
+at this time the values seem more meaningful than rates, particularly where
+there arent continuous streams of activity. Where latencies are shown, they
+are computed using the counters from the past minute.
+
+Depending on the chart and underlying monitoring system, some charts show
+instantaneous value differences (between samples) while others show
+differences over a sliding window. Either way, the dashboards are not
+necessarily precise numbers but should reflect accurate trends and relative
```
Comment (ttomsu) [inline setup/monitoring/index.md]: The phrase "not necessarily precise" when referring to metrics directly gathered from the system gave me pause. It makes the user beg the questions "Why aren't they precise? Is there a way to make them more precise? How do I get more precise numbers?"

Perhaps what you're meaning is that no metric system can be 100% accurate, but that the precision of the absolute values aren't as important as the relative differences over time.

**Review by ttomsu — APPROVED** [review]
Still reading through the individual platform docs, but here are some comments to help get started. Feel free to submit if I don't get to the other docs in time.

File: setup/monitoring/prometheus.md [hunk 18]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+
+The current Spinnaker installation script can configure for a simple
+single instance VM deployment typically used for prototype purposes, or
+a multi-VM "high-availability" production type deployment on Google Cloud
+Platform. For more information and custom configuration options, see the
+[Prometheus Configuration](https://prometheus.io/docs/operating/configuration/)
+documentation. They describe support for discovering services on EC2, Azure, 
+and other platforms or discovery mechanisms you might be using.
+
+Alternitavely, you can deploy Prometheus to use a Gateway server wherein
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md]: Having diagrams illustrating the differences of how data is flowing would be very helpful.

File: setup/monitoring/prometheus.md [hunk 19]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+(which are typically not known and might be in autoscaled instance groups).
+However, Prometheus will still be polling the Gateway server, so the data
+it does receive wont be fresh. Making matters worse, the data will be
+timestamped at the point it is retrieved from the Gateway server, rather
+than at the time it was pushed to the Gateway server. So the data timeline
+might be off by as much as a minute from actual time making it hard to
+correlate monitoring data to logs or other evidence when diagnosing problems
+in addition to the other downsides of having stale data. For long term
+trends, this is not so much of an issue since the integrity of the counters
+and trends they show will still be accurate.
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md:47]: Are you expecting end users to properly weigh this tradeoff? It seems to me that one is clearly preferable - is it necessary to mention this alternative and it's implications?

File: setup/monitoring/prometheus.md [hunk 20]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+
+The Spinnaker installation scripts support Gateway deployments as well.
+However, it is recommended that you use a standard direct polling deployment
+when possible.
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md]: I think this comment further proves that it's not necessary to mention the Gateway option, except maybe in a section much lower in the document

File: setup/monitoring/prometheus.md [hunk 21]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+
+  2. Run the client-side configuration script.
+     ```
+     /opt/spinnaker-monitoring/third_party/prometheus/install \
+        --client_only \
+        --client_host=''
+     ```
+     * The `--client_host` is an optional parameter that configures the
+       Spinnaker Monitoring Daemon to listen on a network interface other
+       than the default `localhost`. The host `` adds all of them. This
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md]: > The host `` adds all of them. 

Suggest rephrasing to "Specify the empty string `''` to add all of them"

File: setup/monitoring/prometheus.md [hunk 22]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+       is illustrated for simplicity. By exposing a network port other
+       than local_host, the server will be able to poll the client from
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md]: local_host ?

File: setup/monitoring/prometheus.md [hunk 23]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+     * You should see the Spinnaker datasource and dashboards from the
+       topleft pulldown. The dashboards may not have any data on them yet
+       because the Daemons might not be polled.
+
+
+# Using The Prometheus Dashboards
+
+Each of the dashboards use Grafana's templating mechanism in order to
+allow you to perform some global filtering. The variables provided depend
+on the type of dashboard. As a rule of thumb, the following are avaialable:
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md]: typo: avaialable

File: setup/monitoring/prometheus.md [hunk 24]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+     between 1 and all, you will need to create your own custom charts using
+     Grafana.
+
+   * There is an application-oriented dashboard that lets you select
+     a particular application, or look at them all. This dashboard only
+     contains metrics that are tagged with application context. Only a
+     few of the metrics within Spinnaker know or care about applications,
+     even though most Spinnaker activity is performed in the context of
+     an application. The list of applications currently comes from a
+     particular metric in Front50. If the application you want does not
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md:212]: > If the application you want does not show up there, you may need to wait a little while longer

Two things with this that may be causing the delay: either 1.) the application is not configured (the config not saved through Front50) or 2.) no operation has been performed on that application, meaning no time-series with that label has been created.

File: setup/monitoring/prometheus.md [hunk 25]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+     choices or create your own custom charts in the meantime.
+
+The dashboards manipulate the names of the metrics and labels in order to
+provide cleaner and more precise legends and labelings. If you need to know
+exact metric names, then you might need to look inside the view definition
+to reverse engineer it.
+
+## Caveats
+     
+The Grafana dashboards are typically "idelta" or "rates" at 1m intervals.
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md]: typo? "idelta"

File: setup/monitoring/prometheus.md [hunk 26]
```diff
@@ -6,3 +6,235 @@ sidebar:
...
+When computing the delta, it considers the previous value in the series
+for the selected time interval. If there was no value (e.g. this is the first
+value seen), then it considers the delta to be 0 rather than the count.
+Unfortunately, that means that the charts will show 0 rather than the actual
+count. This could mean that you might not see errors and other rare anomolies
```
Comment (ttomsu) [inline setup/monitoring/prometheus.md:233]: I remember you telling me about this. It seems like a pretty big issue - Do you know of a way we can tell prometheus to assume 0 if no metric appeared before the first instance?

File: setup/monitoring/stackdriver.md [hunk 27]
```diff
@@ -6,3 +6,92 @@ sidebar:
...
+
+Note that Spinnaker talks to Stackdriver via HTTP so you do not need
+to install the Stackdriver Agent. Since the Agent is only available on
+Google Cloud Platform or Amazon Web Services, this means that you can
+write metrics into Stackdriver regardless of what platform you are running
+Spinnaker on. You may still want to install a Stackdriver Agent for
+local machine monitoring or monitoring other services. However, the
+agent itself does not know about Spinnaker and will not monitor it.
+
+*__Warning__: The current integration with Stackdriver conflicts with
```
Comment (ttomsu) [inline setup/monitoring/stackdriver.md]: While true, I highly suggest rewording this to be less...incendiary. Consider:

"The current integration uses Stackdriver custom metrics, which are the most flexible way to integrate the dynamic number of time-series generated by each of Spinnaker's microservices. As a result, even a single VM deployment can cause hundreds of dollars of charges per month. If you are interested in using Stackdriver, please contact us through the Spinnaker Slack channel"

@duftler

File: setup/monitoring/stackdriver.md [hunk 28]
```diff
@@ -6,3 +6,92 @@ sidebar:
...
+enabled | true or false | Whether or not to write to stackdriver directly
+projectName | Stackdriver Project name | The stackdriver project to write the metric data into. The default is the `providers.google.primaryCredentials.project` however this is is not necessarily correct. Especially if you are using Stackdriver but not using Google.
+credentialsPath | Path to json file | A path to the JSON file containing the service account credentials downloaded from the Google Developer's Console used to authenticate with Stackdriver. The default is providers.google.primaryCredentials.jsonPath, however this might not be the account or credentials you wish to use. If you leave the value empty, it will use the default Service Account credentials for the VM that you are deployed on.
+
+If you are on Google Cloud Platform and using the Default Service Account
+credentials then you may need to enable the Monitoring-Write scope.
+This scope is enabled by default, however depending on how you create
+the VMs that you are deploying into, the scope may have been left out.
+
+Proceed to [install the dashboards](#installing-the-stackdriver-dashboards)
```
Comment (ttomsu) [inline setup/monitoring/stackdriver.md:51]: Probably not necessary to instruct the user to continue to the next section, which is directly below the current position.

File: setup/monitoring/stackdriver.md [hunk 29]
```diff
@@ -6,3 +6,92 @@ sidebar:
...
+and select your project. You should be able to see the various Spinnaker
+dashboards and select from them.
+
+
+## Caveats
+
+Stackdriver dashboards do not have the ability to manipulate metrics so
+can only show raw values coming from Spinnaker. In particular, latencies
+cannot be computed and time values will be in nanoseconds rather than
+milliseconds or seconds.
```
Comment (ttomsu) [inline setup/monitoring/stackdriver.md:84]: Can this conversion be done in the daemon?

**Review by ttomsu — APPROVED** [review]
